### PR TITLE
fix fallback URL

### LINKED ISSUE #456: fallbackSCM no longer honoured agter '404
### Jenkins and plugins versions report

<details>
  <summary>Post #404 the fallback repository no longer works</summary>

  
  ```text
21:05:16  #############################################
21:05:16  #############################################
21:05:16  ##
21:05:16  ## Starting to test redacted_plugin_id redacted_plugin_tag against MavenCoordinates[groupId=org.jenkins-ci.plugins, artifactId=plugin, version=2.375.3-cb-1]
21:05:16  ##
21:05:16  #############################################
21:05:16  #############################################
21:05:16  
21:05:16  
21:05:16  
21:05:16  
21:05:16  
21:05:16  2023-02-10 21:05:16.147+0000 [id=1]	INFO	o.j.t.test.PluginCompatTester#testPluginAgainst: Created plugin checkout directory /jenkins/workspace/builders_URR-pr-builder_PR-8512/output-redacted_plugin_id/work/redacted_plugin_id
21:05:16  2023-02-10 21:05:16.148+0000 [id=1]	INFO	o.j.t.test.PluginCompatTester#getScmTag: Using SCM tag d6b8042a06de4aaaf0942ad79036095b853eea02 from POM
21:05:16  2023-02-10 21:05:16.150+0000 [id=1]	INFO	o.j.t.test.PluginCompatTester#clone: Checking out from SCM connection URL scm:git:https://github.com/jenkinsci/redacted_plugin_id-plugin.git at d6b8042a06de4aaaf0942ad79036095b853eea02
21:05:16  2023-02-10 21:05:16.387+0000 [id=1]	INFO	o.j.t.test.PluginCompatTester#clone: Checking out from SCM connection URL scm:git:git@github.com:jenkinsci-cert/redacted_plugin_id-plugin.git at d6b8042a06de4aaaf0942ad79036095b853eea02
21:05:17  2023-02-10 21:05:17.296+0000 [id=1]	SEVERE	o.j.t.test.PluginCompatTester#testPluginAgainst: Failed to check out plugin sources
21:05:17  java.io.IOException: git fetch origin failed with exit status 128: fatal: remote error: upload-pack: not our ref d6b8042a06de4aaaf0942ad79036095b853eea02
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.clone(PluginCompatTester.java:708)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.cloneFromSCM(PluginCompatTester.java:612)
21:05:17  Caused: java.io.UncheckedIOException
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.cloneFromSCM(PluginCompatTester.java:623)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.testPluginAgainst(PluginCompatTester.java:491)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.testPlugins(PluginCompatTester.java:318)
21:05:17  	at org.jenkins.tools.test.PluginCompatTesterCli.main(PluginCompatTesterCli.java:199)
21:05:17  2023-02-10 21:05:17.298+0000 [id=1]	SEVERE	o.j.t.test.PluginCompatTester#testPlugins: Internal error while executing a test for core 2.375.3-cb-1 and plugin redacted_plugin_id redacted_plugin_tag. Please submit a bug to plugin-compat-tester
21:05:17  java.io.IOException: git fetch origin failed with exit status 128: fatal: remote error: upload-pack: not our ref d6b8042a06de4aaaf0942ad79036095b853eea02
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.clone(PluginCompatTester.java:708)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.cloneFromSCM(PluginCompatTester.java:612)
21:05:17  Caused: java.io.UncheckedIOException
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.cloneFromSCM(PluginCompatTester.java:623)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.testPluginAgainst(PluginCompatTester.java:491)
21:05:17  Caused: org.jenkins.tools.test.exception.PluginSourcesUnavailableException: Failed to check out plugin sources
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.testPluginAgainst(PluginCompatTester.java:505)
21:05:17  	at org.jenkins.tools.test.PluginCompatTester.testPlugins(PluginCompatTester.java:318)
21:05:17  	at org.jenkins.tools.test.PluginCompatTesterCli.main(PluginCompatTesterCli.java:199)
21:05:17  [namaste]: PCT execution finished!
21:05:17  [namaste]: Report available in: /jenkins/workspace/builders_URR-pr-builder_PR-8512/output-redacted_plugin_id/out/report.xml
21:05:17  [namaste]: Finished in: 13s
21:05:17  2023-02-10 21:05:17,821 INFO  [io.quarkus] (main) namaste stopped in 0.022s  
```

The tag and ref does exist in the `jenkins-cert` directory.

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

Linux but probably all.

### Reproduction steps

1. create an war containing staged (but not yet public) security releases.
2.run the pct for a specific plugin that has security fixes

### Expected Results

PCT runs surefire tests for the plugin

### Actual Results

clone/checkout fails and no tests are run

### Anything else?

could probably be tested using a forked repository with a tag/release (without being pushed anywhere)

### PR DESCRIPTION
**Pull request by jtnord**
if the fallbackURL succeeded the exception of any prior attempts was still thrown.

fixes #456 
amends #404 

tested locally with an upcoming security release using the jenkinsci-cert repo. (plugin name, tag and sha redacted)

```
? java -jar target\plugins-compat-tester-cli.jar -fallbackGitHubOrganization jenkinsci-cert  -workDirectory work -includePlugins theplugin -war C:\Users\jnord\Desktop\SECURITY_com_cloudbees_core_traditional_client-master_2.375.3.6-SNAPSHOT_client-master-2.375.3.6-SNAPSHOT.war
2023-02-13 12:22:48.242+0000 [id=1]     INFO    o.j.t.t.h.MultiParentCompileHook#<init>: Loaded multi-parent compile hook
2023-02-13 12:22:48.403+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#scanWAR: Scanned contents of C:\Users\jnord\Desktop\SECURITY_com_cloudbees_core_traditional_client-master_2.375.3.6-SNAPSHOT_client-master-2.375.3.6-SNAPSHOT.war with 220 plugins
2023-02-13 12:22:48.483+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#scanWAR: Scanned contents of C:\Users\jnord\Desktop\SECURITY_com_cloudbees_core_traditional_client-master_2.375.3.6-SNAPSHOT_client-master-2.375.3.6-SNAPSHOT.war with 220 plugins
2023-02-13 12:22:48.489+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#testPlugins: Starting plugin tests on core coordinates MavenCoordinates[groupId=org.jenkins-ci.main, artifactId=jenkins-war, version=2.375.3-cb-1]
2023-02-13 12:22:48.540+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#testPluginAgainst:





#############################################
#############################################
##
## Starting to test theplugin the_plugin_version against MavenCoordinates[groupId=org.jenkins-ci.main, artifactId=jenkins-war, version=2.375.3-cb-1]
##
#############################################
#############################################





2023-02-13 12:22:48.577+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#testPluginAgainst: Created plugin checkout directory C:\workarea\source\github\jenkinsci\plugin-compat-tester\work\theplugin
2023-02-13 12:22:48.577+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#getScmTag: Using SCM tag thecommitsha from POM
2023-02-13 12:22:48.587+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#cloneImpl: Checking out from SCM connection URL https://github.com/jenkinsci/theplugin-plugin.git at thecommitsha
2023-02-13 12:22:52.684+0000 [id=1]     INFO    o.j.t.test.PluginCompatTester#cloneImpl: Checking out from SCM connection URL git@github.com:jenkinsci-cert/theplugin-plugin.git at thecommitsha
2023-02-13 12:22:59.047+0000 [id=1]     INFO    o.j.t.t.m.h.PluginCompatTesterHooks#runHooks: Running hook: org.jenkins.tools.test.hook.NodeCleanupBeforeCompileHook
2023-02-13 12:22:59.048+0000 [id=1]     INFO    o.j.t.t.h.NodeCleanupBeforeCompileHook#action: Hook not triggered; continuing
2023-02-13 12:22:59.051+0000 [id=1]     INFO    o.j.t.t.m.ExternalMavenRunner#run: Running mvn.cmd --show-version --batch-mode --errors -ntp --define=failIfNoTests=false clean process-test-classes -Dmaven.javadoc.skip in C:\workarea\source\github\jenkinsci\plugin-compat-tester\work\theplugin >> C:\workarea\source\github\jenkinsci\plugin-compat-tester\work\logs\theplugin\plugintag_against_org.jenkins-ci.main_jenkins-war_2.375.3-cb-1.log
Apache Maven 3.9.0 (9b58d2bad23a66be161c4664ef21ce219c2c8584)
Maven home: c:\java\maven-3.9.0
Java version: 11.0.15, vendor: Eclipse Adoptium, runtime:
```

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 1]
```diff
@@ -440,12 +440,15 @@ public static void cloneFromScm(
...
+                if (StringUtils.startsWith(connectionURL, "scm:git:")) {
+                    connectionURL = StringUtils.substringAfter(connectionURL, "scm:git:");
+                }
                 // See: https://github.blog/2021-09-01-improving-git-protocol-security-github/
                 connectionURL = connectionURL.replace("git://", "https://");
             }
             try {
                 cloneImpl(connectionURL, scmTag, checkoutDirectory);
-                break;
+                return; // checkout was ok
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:451]: also fixes a bug where if the first clone failed, and subsequent ones passed an exception would still be thrown as `lastException` would not be `null` further down when the loop exited.

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 2]
```diff
@@ -440,12 +440,15 @@ public static void cloneFromScm(
         PluginSourcesUnavailableException lastException = null;
         for (String connectionURL : connectionURLs) {
             if (connectionURL != null) {
+                if (StringUtils.startsWith(connectionURL, "scm:git:")) {
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:443]: code was moved up here so that `cloneImpl` only has to deal with real `git` URLs

**Comment by jtnord** [comment]
https://github.com/isaacs/github/issues/436 (obsolete)

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 3]
```diff
@@ -528,17 +531,21 @@ private static void cloneImpl(String connectionURL, String scmTag, File checkout
...
+        // refs/heads/*:refs/origin/heads/* refs/pull/*/heads:refs/origin/pull/*/heads
         p =
                 new ProcessBuilder()
                         .directory(checkoutDirectory)
-                        .command("git", "fetch", gitUrl, scmTag)
+                        .command(
+                                "git",
+                                "fetch",
+                                gitUrl,
+                                "refs/tags/*:refs/origin/tags/*", // actual releases
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:546]: everything is origin.  it doesn't really matter, but when manually testing running git commands in isolation I got a conflict where `master` was not `master`  (should not happen - but we can be safe).

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 4]
```diff
@@ -472,19 +475,19 @@ public static void cloneFromScm(
      *   <li><code>git checkout FETCH_HEAD</code>
      * </ul>
      *
-     * @param connectionURL The connection URL, in a format such as
-     *     scm:git:https://github.com/jenkinsci/mailer-plugin.git or
-     *     https://github.com/jenkinsci/mailer-plugin.git
+     * @param gitURL The git native URL, see the <a
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:478]: makes the logging easier to follow if it spits out the URL that is actually going to be used for the `git fetch`

**Review by basil — APPROVED** [review]
Whoops, sorry about that! Thanks for the quick fix.

**Pull request merged**